// tradex-common/build/src/modules/errors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const FieldRequiredError_1 = require("./FieldRequiredError");
exports.FieldRequiredError = FieldRequiredError_1.default;
const GeneralError_1 = require("./GeneralError");
exports.GeneralError = GeneralError_1.default;
exports.createFromStatus = GeneralError_1.createFromStatus;
const UriNotFound_1 = require("./UriNotFound");
exports.UriNotFound = UriNotFound_1.default;
const SystemError_1 = require("./SystemError");
exports.SystemError = SystemError_1.default;
const ObjectNotFoundError_1 = require("./ObjectNotFoundError");
exports.ObjectNotFoundError = ObjectNotFoundError_1.default;
const AlreadyExistedError_1 = require("./AlreadyExistedError");
exports.AlreadyExistedError = AlreadyExistedError_1.default;
const InvalidParameterError_1 = require("./InvalidParameterError");
exports.InvalidParameterError = InvalidParameterError_1.default;
const InvalidIdSecretError_1 = require("./InvalidIdSecretError");
exports.InvalidIdSecretError = InvalidIdSecretError_1.default;
const InvalidFieldValueError_1 = require("./InvalidFieldValueError");
exports.InvalidFieldValueError = InvalidFieldValueError_1.default;
const ForwardError_1 = require("./ForwardError");
exports.ForwardError = ForwardError_1.ForwardError;
const TokenExpiredError_1 = require("./TokenExpiredError");
exports.TokenExpiredError = TokenExpiredError_1.default;
const TimeoutError_1 = require("./TimeoutError");
exports.TimeoutError = TimeoutError_1.TimeoutError;
exports.EMAIL_VALIDATION_FAILED = 'EMAIL_VALIDATION_ERROR';
exports.TEMPLATE_LOAD_FAILED = 'TEMPLATE_LOAD_FAILED';
